operator/v1alpha1/types_externaldns.go

```diff
@@ -0,0 +1,397 @@
+	// created if multiple ExternalDNS source resources
+	// are desired.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	Source ExternalDNSSource `json:"source"`
+
+	// Zones describes which DNS Zone IDs
+	// ExternalDNS should publish records to.
+	//
+	// +optional
+	Zones []string `json:"zones,omitempty"`
+}
+
+// ExternalDNSDomain describes how sets of included
+// or excluded domains are to be constructed.
+type ExternalDNSDomain struct {
+	ExternalDNSDomainUnion `json:",inline"`
+
+	// FilterType marks the Name or Pattern field
+	// as an included or excluded set of domains.
+	//
+	// Note that excluded domains that
+	// have not been included will be
+	// ignored.
+	//
+	// This field accepts the following values:
+	//
+	//  "Include": Include the domain set specified
+	//  by name or pattern.
+	//
+	//  "Exclude": Exclude the domain set specified
+	//  by name or pattern.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	FilterType ExternalDNSFilterType `json:"filterType"`
+}
+
+// ExternalDNSDomainUnion describes optional fields of an ExternalDNSDomain
+// that should be captured.
+// +union
+type ExternalDNSDomainUnion struct {
+	// MatchType specifies the type of match to be performed
+	// by ExternalDNS when determining whether or not to publish DNS
+	// records for a given source resource based on the resource's
+	// requested hostname.
+	//
+	// This field accepts the following values:
+	//
+	//  "Exact": Explicitly match the full domain string
+	//   specified via the name field, including any subdomains
+	//   of the given name.
+	//
+	//  "Pattern": Match potential domains against
+	//  the provided regular expression pattern string.
+	//
+	// +unionDiscriminator
+	// +kubebuilder:validation:Required
+	// +required
+	MatchType DomainMatchType `json:"matchType"`
+
+	// Name is a string representing a single domain
+	// value.
+	//
+	// e.g. my-app.my-cluster-domain.com
+	//
+	// +optional
+	Name *string `json:"names,omitempty"`
+
+	// Pattern is a regular expression used to
+	// match a set of domains.
+	//
+	// +optional
+	Pattern *string `json:"pattern,omitempty"`
+}
+
+// +kubebuilder:validation:Enum=Exact;Pattern
+type DomainMatchType string
+
+const (
+	DomainMatchTypeExact DomainMatchType = "Exact"
+	DomainMatchTypeRegex DomainMatchType = "Pattern"
+)
+
+// +kubebuilder:validation:Enum=Include;Exclude
+type ExternalDNSFilterType string
+
+const (
+	FilterTypeInclude ExternalDNSFilterType = "Include"
+	FilterTypeExclude ExternalDNSFilterType = "Exclude"
+)
+
+// ExternalDNSProvider specifies configuration
+// options for the desired ExternalDNS DNS provider.
+type ExternalDNSProvider struct {
+	// Type describes which DNS provider
+	// ExternalDNS should publish records to.
+	// The following DNS providers are supported:
+	//
+	//  * AWS
+	//  * GCP
+	//  * Azure
+	//  * BlueCat
+	//  * Infoblox
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	Type ExternalDNSProviderType `json:"type"`
```

Comment:
If we decide in the future that we need provider-specific parameters, I don't know whether we can change this to a discriminator in a subsequent release, so it's critical that we make this determinism now.  Do we know for sure that we don't need provider-specific parameters?

Reply:
I cannot say for sure that we don't need provider-specific parameters. I'll change this to a union type.

Reply:
Fixed. Hopefully this change comes in handy later :wink: 
Given the sprawl of ExternalDNS, I suspect we will need to express some provider specific parameters via an API at some point.